### DISCUSSION TITLE
2D Version of Curl Integrator

### DISCUSSION THREAD

**Discussion by markusrenoldner** [mfem/mfem#3635] [Q&A] [answered]
Dear community, 

what is the most straight forward way to implement LinearForm integrators and BilinearForm integrators like
```
mfem::VectorFEBoundaryTangentLFIntegrator()
mfem::MixedCrossProductIntegrator()
```
for 2D meshes?

**Comment by tzanio** [comment]
ping: @mlstowell

**Comment by mlstowell** [comment]
Hi, @markusrenoldner,

The 2D version of  `mfem::MixedCrossProductIntegrator()` is the `mfem::MixedScalarCrossProductIntegrator()`. If that isn't exactly what you need you may be able to modify it to suit your purposes. Note that related integrators `mfem::MixedScalarCurlIntegrator()` and `mfem::MixedScalarWeakCurlIntegrator()` may also be of interest.

I'm afraid we don't yet have any detailed guidance on writing your own integrators. However, I'd be happy to offer advice. Specifically which operators/integrators do you need?

Best wishes,
Mark

**Reply by markusrenoldner** [reply]
Hello!

I see, so what I am trying to implement is a 2D version of this BilinearForm:
$$\left\langle u \times v, w\right\rangle \qquad \forall w$$

which corresponds to the `MixedCrossProductIntegrator`. Moreover, I believe that the boundary integral
$$(u\times n, w)_{\partial \Omega}\qquad \forall w$$

implemented by `VectorFEBoundaryTangentLFIntegrator` also only works in 3D and I am looking for a 2D replica as well.

Interestingly, i found that `MixedVectorCurlIntegrator` seems to work in 2D, this I did not expect due to the curl operator.

**Comment by mlstowell** [comment]
Hi, @markusrenoldner,

Just to be clear, what spaces are `u` and `w` meant to be in?

Best wishes,
Mark

**Reply by markusrenoldner** [reply]
Hello!
I am actually interested in both $u, v, w\in H(curl)$ as well as $u, v, w\in H(div)$ for the `MixedCrossProductIntegrator`. (In the discrete setting these terms should all be square integrable).

For the `VectorFEBoundaryTangentLFIntegrator`, I have $u\in H(div)$ and $w\in H(curl)$.

**Comment by mlstowell** [comment, marked as answer]
Hi, @markusrenoldner,

I'm afraid what you're asking for is inconsistent. If `u` and `v` are in HCurl in 2D then their cross product will be a scalar so `w` cannot be in HCurl. On the other hand `u` and `w` could be in HCurl if `v` were scalar and interpreted as a vector perpendicular to the 2D surface. Unless you're planning to use the ND_R2D finite elements which can represent 3 component vector fields in 2D surfaces. If the later is your intention then this will require special attention since many of our bilinear integrators were written before these ND_R2D basis functions were implemented.

Best wishes,
Mark

**Reply by markusrenoldner** [reply]
Good morning,

you are right, I would require $v$ (or $u$) to be a vector perpendicular to the 2D domain. It sounds as if this would be quite some work to implement, which I dont have time for at the moment.

Still, thank you for the response!
